lib/SILOptimizer/Analysis/ProtocolConformanceAnalysis.cpp

```diff
@@ -67,4 +67,51 @@ void ProtocolConformanceAnalysis::init() {
   }
 }
 
+/// Recursively traverse the conformance lists to determine sole conforming
+/// class, struct or enum type.
+NominalTypeDecl *
+ProtocolConformanceAnalysis::findSoleConformingType(ProtocolDecl *Protocol) {
+  SmallVector<ProtocolDecl *, 8> PDWorkList;
+  SmallPtrSet<ProtocolDecl *, 8> VisitedPDs;
+  NominalTypeDecl *SoleConformingNTD = nullptr;
+  PDWorkList.push_back(Protocol);
+  while (!PDWorkList.empty()) {
+    auto *PD = PDWorkList.pop_back_val();
+    // Protocols must have internal or lower access scopes.
+    if (PD->getEffectiveAccess() > AccessLevel::Internal) {
+      return nullptr;
+    }
+    VisitedPDs.insert(PD);
+    auto NTDList = getConformances(PD);
+    if (NTDList.empty())
```

Comment:
I don't think this check is necessary -- a for loop over an empty collection won't do much work.